Introduce authentication for realm-server

FadhlanR opened the pull request:
I moved the `createSession` function from the `Realm` class to a new class called `RealmAuthServer` (I'm open to better naming suggestions). The `RealmAuthServer` is now used in both the `Realm` and `RealmServer` classes. I use `/session` for Realm server authentication to avoid path overlap, as we used `/_session` for Realm authentication.

FadhlanR committed:
Introduce authentication for realm-server

FadhlanR committed:
Merge branch 'main' into cs-7029-realm-server-authentication

FadhlanR committed:
lint fixes

FadhlanR committed:
fix realm tests

FadhlanR committed:
Update scripts

FadhlanR committed:
Rename env vars

habdelra commented:
using `/session` for the realm server endpoint will mean that we need to make this a reserved word when it comes time to creating user accounts. eventually the top path will be the username for a user's personal realm. Probably we should pair this with work to prevent users from making a username called "session" in our create user logic (or if we choose a different word--whatever that word is).

habdelra commented on packages/realm-server/main.ts:
```diff
@@ -43,6 +44,38 @@ if (!REALM_SECRET_SEED) {
   process.exit(-1);
 }
 
+const REALM_SERVER_SECRET_SEED = process.env.REALM_SERVER_SECRET_SEED;
```
it might be easier to just use the REALM_SECRET_SEED instead of making a new seed just for this. since these are encrypted values in our SSM, i would guess that if one could be compromised all of them could be compromised. so making separate ones seems like more of a hassle to me.

habdelra commented on packages/realm-server/scripts/start-base-root.sh:
```diff
@@ -9,6 +9,10 @@ NODE_ENV=development \
   PGPORT=5435 \
   PGDATABASE=boxel_test_base_root \
   REALM_SECRET_SEED="shhh! it's a secret" \
+  REALM_SERVER_SECRET_SEED="shhh! it's a secret" \
```
like i mentioned above, i'm not totally convinced having 2 seeds makes things more secure, and probably just extra hassle

habdelra commented:
Is this PR paired with an infra PR to set the new env vars? if not could you make that as well?

FadhlanR commented:
> Is this PR paired with an infra PR to set the new env vars? if not could you make that as well?

I just submitted https://github.com/cardstack/infra/pull/537, I'll remove the REALM_SERVER_SECRET_SEED.

FadhlanR committed:
Remove REALM_SERVER_SECRET_SEED

habdelra commented on packages/runtime-common/realm-auth-server.ts:
I like this abstraction--good job!

habdelra commented:
Let's also update the user creation so that "session" is a forbidden username

FadhlanR committed:
claim session username

FadhlanR committed:
Update comment

FadhlanR committed:
Merge branch 'main' into cs-7029-realm-server-authentication

lukemelia commented:
Naming suggestions:

RealmAuthServer -> MatrixBackedAuthentication
/session -> /_server-session

habdelra commented:
Fadhlan and I discussed, and we thought the best approach for preserving the /session (or /_server-session) namespace was to create a script that would register a matrix user with this name and a random password so that users couldn't claim this name.

although, if we are going with our convention, `_` prefix is probably off limits for users. this guards against other endpoints too actually....

habdelra commented:
@FadhlanR lets hold off on any approach around reserving usernames until tomorrows (thur) talk.

FadhlanR committed:
Merge branch 'main' into cs-7029-realm-server-authentication

FadhlanR committed:
Rename to _server-session and MatrixBackendAuthentication

FadhlanR commented:
Since we decided to use usernames as part of realm server routes, I think it’s better to treat certain endpoints, like session or realm endpoints, differently. Following @lukemelia 's suggestion, I renamed the endpoint to `_server-session`. I also updated how we reserve words for endpoints. Instead of registering users to reserve ‘session,’ I added validation in the user registration flow. Please check this commit [5a29a03](https://github.com/cardstack/boxel/pull/1485/commits/5a29a036d5e4572ea2d94b2f42763e400a5173d0).
